src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -170,8 +171,16 @@ internal static bool AddExtraChainCertificates(SafeSslHandle sslContext, X509Cha
             Debug.Assert(chain != null, "X509Chain should not be null");
             Debug.Assert(chain.ChainElements.Count > 0, "chain.Build should have already been called");
 
-            // Don't count the last item (the root)
-            int stop = chain.ChainElements.Count - 1;
+            // If the last certificate is a root certificate, don't send it. PartialChain means the last cert wasn't a root.
+            int stop;
+            if (chain.ChainStatus.Any(s => (s.Status & X509ChainStatusFlags.PartialChain) != 0))
+            {
+                stop = chain.ChainElements.Count;
+            }
+            else
+            {
+                stop = chain.ChainElements.Count - 1;
+            }
```

Comment:
Can we instead do:
```C#
int stop = chain.ChainElements.Count;
foreach (X509ChainStatus s in chain.ChainStatus)
{
    if ((s.Status & X509ChainStatusFlags.PartialChain) != 0)
    {
        stop--;
        break;
    }
}
```
That should save an allocation, avoid interface and delegate calls, and avoid the need to add the LINQ dependency.

Reply:
Ok.  I can do that. I was under the impression that Linq was efficient at the compiler level. Is that not true?  Is it discouraged to use in product code?

Reply:
> Is that not true?

The C# compiler does not optimize it.

> Is it discouraged to use in product code?

In our core libraries where perf can be a factor, yes.